airflow/jobs/scheduler_job_runner.py

```diff
@@ -1919,23 +1919,27 @@ def adopt_or_reset_orphaned_tasks(self, session: Session = NEW_SESSION) -> int:
         return len(to_reset)
 
     @provide_session
-    def check_trigger_timeouts(self, session: Session = NEW_SESSION) -> None:
+    def check_trigger_timeouts(
+        self, max_retries: int = MAX_DB_RETRIES, session: Session = NEW_SESSION
+    ) -> None:
         """Mark any "deferred" task as failed if the trigger or execution timeout has passed."""
-        num_timed_out_tasks = session.execute(
-            update(TI)
-            .where(
-                TI.state == TaskInstanceState.DEFERRED,
-                TI.trigger_timeout < timezone.utcnow(),
-            )
-            .values(
-                state=TaskInstanceState.SCHEDULED,
-                next_method="__fail__",
-                next_kwargs={"error": "Trigger/execution timeout"},
-                trigger_id=None,
-            )
-        ).rowcount
-        if num_timed_out_tasks:
-            self.log.info("Timed out %i deferred tasks without fired triggers", num_timed_out_tasks)
+        for attempt in run_with_db_retries(max_retries, logger=self.log):
```

Comment:
Instead of implementing q retry loop in the function body, have you considered implementing the retry via a `@tenacity.retry()` decorator? That makes to code nicer and lean.

Reply:
Hi @jscheffl Thanks for the review. I understand your suggestion, but this is a special retry mechanism for DB-related errors with specifically designed retry intervals.
If I use tenacity.retry, I would need to include the retry_kwargs in the decorator, which might not look as clean or neat. Additionally, I want to align with the current approach for retrying on DB issues, for example:

https://github.com/apache/airflow/blob/main/airflow/jobs/scheduler_job_runner.py#L1811
https://github.com/apache/airflow/blob/main/airflow/models/trigger.py#L165

<img width="885" alt="Screenshot 2024-10-01 at 12 59 33" src="https://github.com/user-attachments/assets/ba8c355c-4cf8-4c20-a1ab-108d7c99e594">

Reply:
That sounds reasonable, thanks.